lib/src/protocol/kawa_h1/mod.rs

```diff
@@ -0,0 +1,1933 @@
+                backend.failures = 0;
+                backend.active_requests += 1;
+                backend.retry_policy.succeed();
+            }
+        }
+    }
+
+    fn fail_backend_connection(&mut self, metrics: &SessionMetrics) {
+        if let Some(backend) = &self.backend {
+            let mut backend = backend.borrow_mut();
+            backend.failures += 1;
+
+            let already_unavailable = backend.retry_policy.is_down();
+            backend.retry_policy.fail();
+            incr!(
+                "connections.error",
+                self.cluster_id.as_deref(),
+                metrics.backend_id.as_deref()
+            );
+            if !already_unavailable && backend.retry_policy.is_down() {
+                error!(
+                    "backend server {} at {} is down",
+                    backend.backend_id, backend.address
+                );
+                incr!(
+                    "down",
+                    self.cluster_id.as_deref(),
+                    metrics.backend_id.as_deref()
+                );
+
+                push_event(Event {
+                    kind: EventKind::BackendDown as i32,
+                    backend_id: Some(backend.backend_id.to_owned()),
+                    address: Some(backend.address.to_string()),
+                    cluster_id: None,
+                });
+            }
+        }
+    }
+
+    pub fn backend_hup(&mut self) -> StateResult {
+        // println!("=====backend_hup");
+        if self.backend_readiness.event.is_readable()
+            && self.backend_readiness.interest.is_readable()
+        {
+            // there might still data we can read on the socket
+            // println!("=====backend_hup:continue");
+            return StateResult::Continue;
+        }
+        if self.response_stream.is_terminated() {
+            // the backend finished to answer we can close
+            // println!("=====backend_hup:terminated");
+            return StateResult::CloseBackend;
+        }
+        match (
+            self.request_stream.is_initial(),
+            self.response_stream.is_initial(),
+        ) {
+            // the backend started to answer so we close
+            (_, false) => {
+                // println!("=====backend_hup:close");
+                StateResult::CloseSession
+            }
+            // probably backend hup between keep alive request, change backend
+            (true, true) => {
+                // println!("=====backend_hup:close_backend");
+                StateResult::CloseBackend
+            }
+            // the frontend already transmitted data so we can't redirect
+            (false, true) => {
+                // println!("=====backend_hup:503");
+                self.set_answer(DefaultAnswerStatus::Answer503, None);
+                self.backend_readiness.interest = Ready::EMPTY;
+                StateResult::Continue
+            }
+        }
+    }
+
+    fn ready_inner(
```

Comment:
please write a documenting comment, like "this is where the proxy magic happens and where traffic is transmitted from the front to the back"